// ngx-capture/esm2020/lib/ngx-capture.module.mjs
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NgxCaptureComponent } from './ngx-capture.component';
import * as i0 from "@angular/core";
export class NgxCaptureModule {
}
NgxCaptureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgxCaptureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxCaptureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.8", ngImport: i0, type: NgxCaptureModule, declarations: [NgxCaptureComponent], exports: [NgxCaptureComponent] });
NgxCaptureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgxCaptureModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgxCaptureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxCaptureComponent],
                    imports: [],
                    exports: [NgxCaptureComponent],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA]
                }]
        }] });
