// ngx-capture/esm2020/lib/ngx-capture.service.mjs
import { Injectable } from '@angular/core';
import html2canvas from 'html2canvas';
import { from, of } from 'rxjs';
import * as i0 from "@angular/core";
export class NgxCaptureService {
    getImage(screen, fullCapture, cropDimensions) {
        let options = {
            logging: false,
            useCORS: true,
            scale: 1
        };
        if (!fullCapture && cropDimensions.width > 10 && cropDimensions.height > 10) {
            options = { ...options, ...cropDimensions };
        }
        else if (!fullCapture) {
            return of(null);
        }
        return from(html2canvas(screen, options)
            .then((canv) => {
            return canv.toDataURL('image/png');
        }, (err) => {
            throw new Error(err);
        })
            .catch((res) => {
            throw new Error(res);
        }));
    }
    downloadImage(img) {
        const element = document.createElement('a');
        element.setAttribute('href', img);
        element.setAttribute('download', 'capture');
        element.style.display = 'none';
        document.body.appendChild(element);
        element.click();
        document.body.removeChild(element);
    }
}
NgxCaptureService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgxCaptureService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NgxCaptureService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgxCaptureService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.8", ngImport: i0, type: NgxCaptureService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
